### ISSUE TITLE
Accessibility: proposal to consider changing the HTML output of $BBOLD be <strong>

### ISSUE THREAD

**Issue by taniwallach** [openwebwork/pg#416]
As far as I understand, WCAG 2.0 accessibility compliance recommends/requires using the HTML `<strong>` tag to generate "strong emphasis" by a screen reader when necessary, while the older `<b>` tag for bold does not provide any screen-reader hint about emphasis.

Although it is not 100% true, much author use of **bold** will be to emphasize visually with an implicit "semantic" meaning. 

As such, I would like to recommend considering whether the existing  `$BBOLD` (and the corresponding `$EBOLD`) be changed to use `<strong>` instead of `<b>` by default.

Alternately, we could add `$BSTRONG` and `$ESTRONG` and recommend that problem authors make changes on their own.

If HTML `strong` does become the default interpretation of `$BBOLD` we could add a  `$BJUSTBOLD` or the like for when plain bold (`<B>`) is the "right thing" in HTML output.

To some extent, the same question may apply to `$BITALIC` giving HTML `<em>` instead of `<i>`, but I suspect that much "mathematical use" of italics is not for emphasis but to set of special terms, so that change might be less clear than that for bold.

@Alex-Jordan - you seem to be the local accessibility expert or have access to accessibility experts - what do you/they say?

---

References:
  - https://accessibility.psu.edu/boldfacehtml/
  - https://www.freecodecamp.org/forum/t/bold-and-strong-text/178376/3
  - https://stackoverflow.com/questions/2243074/is-it-ok-to-use-strong-in-place-of-b-blindly
  - https://stackoverflow.com/questions/2243074/is-it-ok-to-use-strong-in-place-of-b-blindly
  - https://html.spec.whatwg.org/#the-b-element
  - https://thatdevgirl.com/blog/bold-vs-strong
  - https://www.brucelawson.co.uk/2018/screenreader-support-for-text-level-semantics/

**Comment by dlglin** [comment]
It’s worth noting that these constructions are only used in text mode.  Any typeset math would use standard LaTeX constructions to obtain bold or italics.  This further supports Tani’s observation that $BBOLD is primarily used to emphasize visually.  I think that this is a reasonable change.  For any edge cases where bold text is serving some other purpose, we could handle these as they are reported.

On Aug 18, 2019, at 3:52 AM, Nathan Wallach <notifications@github.com<mailto:notifications@github.com>> wrote:


As far as I understand, WCAG 2.0 accessibility compliance recommends/requires using the HTML <strong> tag to generate "strong emphasis" by a screen reader when necessary, while the older <b> tag for bold does not provide any screen-reader hint about emphasis.

Although it is not 100% true, much author use of bold will be to emphasize visually with an implicit "semantic" meaning.

As such, I would like to recommend considering whether the existing $BBOLD (and the corresponding $EBOLD) be changed to use <strong> instead of <b> by default.

Alternately, we could add $BSTRONG and $ESTRONG and recommend that problem authors make changes on their own.

If HTML strong does become the default interpretation of $BBOLD we could add a $BJUSTBOLD or the like for when plain bold (<B>) is the "right thing" in HTML output.

To some extent, the same question may apply to $BITALIC giving HTML <em> instead of <i>, but I suspect that much "mathematical use" of italics is not for emphasis but to set of special terms, so that change might be less clear than that for bold.

@Alex-Jordan<https://github.com/Alex-Jordan> - you seem to be the local accessibility expert or have access to accessibility experts - what do you/they say?

________________________________

References:

  *   https://accessibility.psu.edu/boldfacehtml/
  *   https://www.freecodecamp.org/forum/t/bold-and-strong-text/178376/3
  *   https://stackoverflow.com/questions/2243074/is-it-ok-to-use-strong-in-place-of-b-blindly
  *   https://stackoverflow.com/questions/2243074/is-it-ok-to-use-strong-in-place-of-b-blindly
  *   https://html.spec.whatwg.org/#the-b-element
  *   https://thatdevgirl.com/blog/bold-vs-strong
  *   https://www.brucelawson.co.uk/2018/screenreader-support-for-text-level-semantics/

**Comment by jwj61** [comment]
I think it makes sense to change the default behavior, and see if anyone even wants BJUSTBOLD.

**Comment by heiderich** [comment]
As I understand it, `<strong>` and `<em>` are rather semantic, while `<b>` and `<i>` are about formatting. As an author I would like to be able to decide whether I want to convey semantic information or instructions about formatting. Therefore I would favor the following alternative:

> Alternately, we could add $BSTRONG and $ESTRONG and recommend that problem authors make changes on their own.

In the original proposal (make `$BBOLD` render as `<strong>` instead of `<b>`) something like the following would become necessary to be able to distinguish the two:

> If HTML strong does become the default interpretation of $BBOLD we could add a $BJUSTBOLD or the like for when plain bold (`<B>`) is the "right thing" in HTML output.

I would prefer to write `$BBOLD` when I mean "bold" instead of `$BJUSTBOLD`.

> Although it is not 100% true, much author use of bold will be to emphasize visually with an implicit "semantic" meaning

It would probably not be too hard to replace  `$BBOLD` automatically by `$BSTRONG` in the OPL and later revert the (probably few) instances where only "formatting" is meant. However this may render it incompatible with older versions of pg.

Is there any meta information on which version of pg is required to render a pg problem? This would help to deal with such transitions in the future.

**Comment by mgage** [comment]
On Aug 19, 2019, at 5:21 PM, Florian Heiderich <notifications@github.com<mailto:notifications@github.com>> wrote:

Is there any meta information on which version of pg is required to render a pg problem? This would help to deal with such transitions in the future.


not yet.  Maybe for 2.16

**Comment by mgage** [comment]
Have you thought about how strong and boldface will be differentiated when producing 
tex output? Or would you make them equivalent in that case. I kind of agree with John Jones: make a reasonable choice and then see if there are important corner cases we haven't thought of.  It is pretty easy to redefine $BBOLD within a problem if there is a special case.

**Comment by taniwallach** [comment]
I made https://github.com/openwebwork/pg/pull/426 to implement the change to using `<Strong>` instead of `<B>`.

> Have you thought about how strong and boldface will be differentiated when producing
> tex output? Or would you make them equivalent in that case. I kind of agree with John Jones: make a reasonable choice and then see if there are important corner cases we haven't thought of. It is pretty easy to redefine $BBOLD within a problem if there is a special case.

LaTeX generated PDF output is very "inaccessible". The document fails to have computer readable document structure ("tags") as well as many other issues. Work is being done to make TeX more accessibility friendly - but it is a long term project. For now, the TeX output can just remain the same.

See: 
  - https://tug.org/twg/accessibility/
  - http://web.science.mq.edu.au/~ross/TaggedPDF/
  - http://web.science.mq.edu.au/~ross/TaggedPDF/PDF-standards-v2.pdf
